### PR TITLE
Add support for generated Java code with KSP

### PR DESCRIPTION
**Pull request by zalewskise**
### What has changed?
- Added support for Java generated code with KSP
- Renamed `deshading` to `reshading` to cover the case where plugins dependencies are clashing with kotlin compiler ones (proper solution would be: https://github.com/bazelbuild/rules_kotlin/issues/624) and those need to be shaded.
- Added guava shading rules
- Added dagger code, generated with ksp plugin, to ksp sample project

### Why it was changed?
- To be able to use Dagger’s KSP processor with `kt_ksp_plugin`

### Fixes
- https://github.com/bazelbuild/rules_kotlin/issues/1074
- https://github.com/bazelbuild/rules_kotlin/issues/1138

### CODE REVIEW

File: kotlin/internal/jvm/compile.bzl [hunk 1]
```diff
@@ -772,7 +772,7 @@ def _run_kt_java_builder_actions(
         java_info = java_common.compile(
             ctx,
             source_files = srcs.java,
-            source_jars = generated_kapt_src_jars + srcs.src_jars,
+            source_jars = generated_kapt_src_jars + srcs.src_jars + generated_ksp_src_jars,
```
Comment (Bencodes) [inline kotlin/internal/jvm/compile.bzl:775]: Doesn't this essentially mean that all targets that execute KSP will be forced to run a Java compilation action even if the underlying KSP plugin only produces Kotlin or class file outputs?

File: kotlin/internal/jvm/compile.bzl [hunk 2]
```diff
@@ -760,9 +760,9 @@ def _run_kt_java_builder_actions(
         java_infos.append(kt_java_info)
 
     # Build Java
-    # If there is Java source or KAPT generated Java source compile that Java and fold it into
+    # If there is Java source or KAPT/KSP generated Java source compile that Java and fold it into
     # the final ABI jar. Otherwise just use the KT ABI jar as final ABI jar.
-    if srcs.java or generated_kapt_src_jars or srcs.src_jars:
+    if srcs.java or generated_kapt_src_jars or srcs.src_jars or generated_ksp_src_jars:
```
Comment (Bencodes) [inline kotlin/internal/jvm/compile.bzl:765]: Doesn't this essentially mean that all targets that execute KSP will be forced to run a Java compilation action even if the underlying KSP plugin only produces Kotlin or class file outputs?

Comment (zalewskise) [reply kotlin/internal/jvm/compile.bzl:765]: Great point, that's correct. 

I'm thinking how we could avoid that and I'm not really sure how we could get the information if the plugin is generating java code or not, without getting this provided as one of `kt_ksp_plugin` parameters. 
Could be something like
```python
kt_ksp_plugin = rule(
    attrs = {
        "generates_java": attr.bool(
                doc = """Runs Java compilation action for plugin generating Java output"""",
                default = False,
        ),
       ...
    }
    ...
}    
``` 
What do you think about this approach? Or maybe there is some detail information that I'm missing, that could be used to determine if we could enforce/avoid running Java compilation action for certain KSP plugins? 👀

Comment (zalewskise) [reply kotlin/internal/jvm/compile.bzl]: @Bencodes updated PR with the approach mentioned above

**Review by Bencodes — APPROVED** [review]

**Pull request merged**